tests/functional/gin.cpp

```diff
@@ -0,0 +1,268 @@
+/*
+ * Copyright (c) 2025 Amazon.com, Inc. or its affiliates. All rights reserved.
+ */
+
+#include "config.h"
+
+#include "test-common.h"
+
+#include <deque>
+#include <vector>
+
+#define PROC_NAME_IDX(i) (i * MPI_MAX_PROCESSOR_NAME)
+
+static void get_ext(ncclNet_v10_t **extNet, ncclGin_v11_t **extGin)
+{
+	void *netPluginLib = dlopen("libnccl-net.so", RTLD_NOW | RTLD_LOCAL);
+	if (netPluginLib == NULL) {
+		NCCL_OFI_WARN("Unable to load libnccl-net.so: %s", dlerror());
+		abort();
+	}
+
+	*extNet = (ncclNet_v10_t *)dlsym(netPluginLib, "ncclNetPlugin_v10");
+	if (*extNet == NULL) {
+		NCCL_OFI_WARN("NetPlugin, could not find ncclNetPlugin_v10 symbol");
+		assert(false);
+	}
+
+	*extGin = (ncclGin_v11_t *)dlsym(netPluginLib, "ncclGinPlugin_v11");
+	if (*extGin == NULL) {
+		NCCL_OFI_WARN("NetPlugin, could not find ncclGinPlugin_v11 symbol");
+		assert(false);
+	}
+}
+
+static inline ncclResult_t poll_request_completion(ncclGin_v11_t *extGin, std::deque<void*> &request_deque, void *collComm)
+{
+	/* Wait for outstanding requests */
+	int done = 0;
+	OFINCCLCHECK(extGin->test(collComm, request_deque.front(), &done));
+	if (done) {
+		request_deque.pop_front();
+	} else {
+		OFINCCLCHECK(extGin->ginProgress(collComm));
+	}
+	return ncclSuccess;
+}
+
+
+int main(int argc, char *argv[])
+{
+	ncclResult_t res = ncclSuccess;
+	int rank, nranks, proc_name_len, local_rank = 0;
+	int buffer_type = NCCL_PTR_HOST;
+
+	/* Plugin defines */
+	int ndev;
+
+	int dev;
+
+	/* Start up MPI */
+	MPI_Init(&argc, &argv);
+	MPI_Comm_rank(MPI_COMM_WORLD, &rank);
+	MPI_Comm_size(MPI_COMM_WORLD, &nranks);
+
+	std::vector<char[NCCL_NET_HANDLE_MAXSIZE]> handles(nranks);
+	std::vector<void*> handles_ptrs(nranks);
+
+	ncclNet_v10_t *extNet = nullptr;
+	ncclGin_v11_t *extGin = nullptr;
+
+	ofi_log_function = logger;
+
+	if (nranks < 2) {
+		NCCL_OFI_WARN("Expected at least two ranks but got %d. "
+			"The gin functional test should be run with at least two ranks.",
+			nranks);
+		res = ncclInvalidArgument;
+		return res;
+	}
+
+	/* All processors IDs, used to find out the local rank */
+	std::vector<char> all_proc_name(nranks * MPI_MAX_PROCESSOR_NAME);
+
+	MPI_Get_processor_name(&all_proc_name[PROC_NAME_IDX(rank)], &proc_name_len);
+	MPI_Allgather(MPI_IN_PLACE, 0, MPI_DATATYPE_NULL, all_proc_name.data(),
+		      MPI_MAX_PROCESSOR_NAME, MPI_BYTE, MPI_COMM_WORLD);
+
+	/* Determine local rank */
+	for (int i = 0; i < nranks; i++) {
+		if (!strcmp(&all_proc_name[PROC_NAME_IDX(rank)],
+			    &all_proc_name[PROC_NAME_IDX(i)])) {
+			if (i < rank) {
+				++local_rank;
+			}
+		}
+	}
+
+	/* Set CUDA device for subsequent device memory allocation, in case GDR is used */
```

Comment:
I was expecting a `cudaSetDevice` from reading this comment but I see `NCCL_OFI_TRACE` call.